R/inputs-spectrum.R

```diff
@@ -208,27 +207,73 @@ read_dp_art_dec31 <- function(dp) {
+  ##   "<AdultARTAdjFactorFlag>" was removed from the .DP file.
+  ## * New tag "<AdultPatsAllocToFromOtherRegion>" was added allowing for input
+  ##   of absolute count adjustment
+  ##
+  ## New logic to account for these changes:
+  ## * Initialise values to defaults 1.0 for relative adjustment and 0.0
+  ##   for absolute adjustment.
+  ## * Only check flag variable if it exists. If adjustment variable exists
+  ##   but flag variable does not exist, use the adjustment.
+
+  ## Initialise
+  adult_artadj_factor <- array(1.0, dim(art15plus_num))
+  dimnames(adult_artadj_factor) <- list(sex = c("male", "female"), year = proj.years)
 
-  if (exists_dptag("<AdultARTAdjFactorFlag>") &&
-        dpsub("<AdultARTAdjFactorFlag>", 2, 4) == 1) {
+  adult_artadj_absolute <- array(0.0, dim(art15plus_num))
+  dimnames(adult_artadj_absolute) <- list(sex = c("male", "female"), year = proj.years)
+
+  ## Flag to use adjustment
+  use_artadj <- exists_dptag("<AdultARTAdjFactor>") &&
+    (!exists_dptag("<AdultARTAdjFactorFlag>") ||
+       (exists_dptag("<AdultARTAdjFactorFlag>") &&
+          dpsub("<AdultARTAdjFactorFlag>", 2, 4) == 1))
+
+  if (use_artadj) {
 
     adult_artadj_factor <- sapply(dpsub("<AdultARTAdjFactor>", 3:4, timedat.idx), as.numeric)
 
+    if(exists_dptag("<AdultPatsAllocToFromOtherRegion>")) {
+      adult_artadj_absolute <- sapply(dpsub("<AdultPatsAllocToFromOtherRegion>", 3:4, timedat.idx), as.numeric)
+    }
+
     ## Only apply if is number (! is percentage)
     adult_artadj_factor <- adult_artadj_factor ^ as.numeric(!art15plus_isperc)
+    adult_artadj_absolute <- adult_artadj_absolute * as.numeric(!art15plus_isperc)
+
+    ## First add absolute adjustment, then apply scalar adjustment (Spectrum procedure)
+    art15plus_num_adj <- art15plus_num + adult_artadj_absolute
+    art15plus_num_adj <- art15plus_num * adult_artadj_factor
 
-    art15plus_num <- art15plus_num * adult_artadj_factor
   }
 
+  # Covert percentage coverage to absolute numbers on ART
   art15plus_num[art15plus_isperc == 1] <- art15plus_need[art15plus_isperc == 1] * art15plus_num[art15plus_isperc == 1] / 100
 
+  # Adjusted number on ART
   art15plus <- as.data.frame.table(art15plus_num,
                                    responseName = "art_dec31",
                                    stringsAsFactors = FALSE)
+  # ART adjustment factor
+  art15plus_adjfactor <- as.data.frame.table(adult_artadj_factor,
+                                   responseName = "artadj_factor",
+                                   stringsAsFactors = FALSE)
+  # ART reallocation
+  art15plus_adjabsolute <- as.data.frame.table(adult_artadj_absolute,
+                                             responseName = "artadj_absolute",
+                                             stringsAsFactors = FALSE)
+
+  art15plus$artadj_factor <- art15plus_adjfactor$artadj_factor
+  art15plus$artadj_absolute <- art15plus_adjabsolute$artadj_absolute
```

Comment:
This is probably fine in practice, but relies on the rows of `art15plus_adjfactor` being in the same ordering as `art15plus` (which should be the case based on how they are constructed). But might be slightly more defensive to assign via joining the data frames

Reply:
Updated to left_join